Allow `--no-src` during clones and `git worktree` after clones

derrickstolee opened the pull request:
These are two highly-requested items from an internal team considering a move to Scalar using Azure Repos.

1. Remove the requirement that we create a `src` directory at clone time.
2. Allow `git worktree` even when using the GVFS protocol.

These are not difficult to implement. The `--no-src` option could even be submitted upstream (though the commit will need to drop one bit about an interaction with the local cache path).

derrickstolee committed:
scalar: .scalarCache should live above enlistment

We should not be putting the .scalarCache inside the enlistment as a
sibling to the 'src' directory. This only happens in "unattended" mode,
but it also negates any benefit of a shared object cache because each
enlistment absolutely does not share any objects with others.

Move the shared object cache in this case to a level above the
enlistment, so at least there is some hope that it can be reused. This
is also critical to the upcoming --no-src option, since the shared
object cache cannot be located within the Git repository.

Signed-off-by: Derrick Stolee <derrickstolee@github.com>

derrickstolee committed:
scalar: add --[no-]src option

Some users have strong aversions to Scalar's opinion that the repository
should be in a 'src' directory, even though it creates a clean slate for
placing build outputs in adjacent directories.

The --no-src option allows users to opt-out of the default behavior.

Signed-off-by: Derrick Stolee <derrickstolee@github.com>

derrickstolee committed:
worktree: allow in Scalar repositories

The 'git worktree' command was marked as BLOCK_ON_GVFS_REPO because it
does not interact well with the virtual filesystem of VFS for Git. When
a Scalar clone uses the GVFS protocol, it enables the
GVFS_BLOCK_COMMANDS flag, since commands like 'git gc' do not work well
with the GVFS protocol.

However, 'git worktree' works just fine with the GVFS protocol since it
isn't doing anything special. It copies the sparse-checkout from the
current worktree, so it does not have performance issues.

This is a highly requested option.

The solution is to stop using the BLOCK_ON_GVFS_REPO option and instead
add a special-case check in cmd_worktree() specifically for a particular
bit of the 'core_gvfs' global variable (loaded by very early config
reading) that corresponds to the virtual filesystem. The bit that most
closely resembled this behavior was non-obviously named, but does
provide a signal that we are in a Scalar clone and not a VFS for Git
clone. The error message is copied from git.c, so it will have the same
output as before if a user runs this in a VFS for Git clone.

Signed-off-by: Derrick Stolee <derrickstolee@github.com>

dscho commented on path.h:
```diff
@@ -182,4 +182,6 @@ const char *git_path_shallow(struct repository *r);
 
 int ends_with_path_components(const char *path, const char *components);
 
+void strip_last_component(struct strbuf *path);
```
Since this is no longer file-local but global, I would like to have the word "path" in the function name. How about `strip_last_path_component()`?

dscho commented on scalar.c:
```diff
@@ -825,6 +827,8 @@ static int cmd_clone(int argc, const char **argv)
 		OPT_BOOL(0, "single-branch", &single_branch,
 			 N_("only download metadata for the branch that will "
 			    "be checked out")),
+		OPT_BOOL(0, "src", &src,
```
Since the default is to _have_ a `src/` directory, we should specify the `"no-src"` form here, so that the `-h` output documents the `--no-src` option (compare e.g. to [`git clone`'s `--no-hardlinks` option](https://github.com/git/git/blob/v2.38.0/builtin/clone.c#L112-L113)).

derrickstolee replied:
I keep forgetting that the parse-opt logic can figure out that `OPT_BOOL(0, "no-*", ...)` will allow `--*`.

dscho commented on t/t9099-scalar.sh:
```diff
@@ -269,4 +269,12 @@ test_expect_success '`scalar delete` with existing repo' '
 '
 
+test_expect_success '`scalar clone --no-src`' '
+	scalar clone --src "file://$(pwd)" with-src &&
```
It would probably be a good idea to drop the `--src` here, to document that the default mode _is_ to have a `src/` directory.

derrickstolee replied:
The point of this test is to make sure the `--src` option is available. The other tests check the implicit `--src` option.

dscho commented on gvfs.h:
```diff
@@ -14,7 +14,18 @@
+ * that.
+ */
 #define GVFS_NO_DELETE_OUTSIDE_SPARSECHECKOUT       (1 << 3)
+#define GVFS_USE_VIRTUAL_FILESYSTEM                 (1 << 3)
```
I could imagine that it might be better to rename `GVFS_NO_DELETE_OUTSIDE_SPARSECHECKOUT` to `GVFS_USE_VIRTUAL_FILESYSTEM` in a separate commit (fixing up, and amending the commit message of, 999461530634b67a79ce2027e407b6e6e2a9e5d1).

What do you think?

Speaking about commit messages... This part of the current commit's message only makes sense if the reader happens to have read the parent commit's message:

> This is also a highly requested option.

To age well, this sentence should be augmented to provide some context, e.g. "Just like the `--no-src` option that was added in the previous commit, support for `git worktree` in Scalar is also a highly requested feature."

derrickstolee replied:
The place where we _use_ `GVFS_NO_DELETE_OUTSIDE_SPARSECHECKOUT` is very specific to deleting paths that are marked with `SKIP_WORKTREE`, hence me wanting to keep the name.

derrickstolee replied:
Commit message is updated to just say `This is a highly requested option.`.

dscho reviewed (approved):
Very nice! I am sure that this will be highly appreciated by the existing and future Scalar users.

I just had a couple of suggestions, nothing major. If you mind strongly, I won't object to merging this as-is, of course.

derrickstolee committed:
abspath: make strip_last_path_component() global

The strip_last_component() method is helpful for finding the parent
directory of a path stored in a strbuf. Extract it to a global method
advertised in path.h. With that additional visibility, it is helpful to
rename it to be more specific to paths.

Signed-off-by: Derrick Stolee <derrickstolee@github.com>

derrickstolee commented:
I rebased this onto `dscho:tentative/vfs-2.38.0`, resolving some interesting conflicts, available as `derrickstolee:src-and-worktrees-2.38.0`.

(I also realized that I had forgotten the `--no-src` swap in this PR, but I caught it in that branch.)

derrickstolee commented:
Closing this since we don't want to merge into `vfs-2.37.3` anymore.
